lntemp/harness.go

```diff
@@ -80,3 +93,376 @@ func NewHarnessTest(t *testing.T, lndBinary string,
+		// We now want to mark the parent harness as cleaned to avoid
+		// running cleanup again since its internal state has been
+		// cleaned up by its child harness tests.
+		h.cleaned = true
+	}
+
+	return st, cleanup
+}
+
+// shutdownNonStandbyNodes will shutdown any non-standby nodes.
+func (h *HarnessTest) shutdownNonStandbyNodes() {
+	for pks, node := range h.manager.activeNodes {
+		// If it's a standby node, skip.
+		_, ok := h.manager.standbyNodes[pks]
+		if ok {
+			continue
+		}
+
+		// The process may not be in a state to always shutdown
+		// immediately, so we'll retry up to a hard limit to ensure we
+		// eventually shutdown.
+		err := wait.NoError(func() error {
+			return h.manager.shutdownNode(node)
+		}, DefaultTimeout)
+		require.NoErrorf(h, err, "unable to shutdown %s", node.Name())
+	}
+}
+
+// cleanupStandbyNode is a function should be called with defer whenever a
+// subtest is created. It will reset the standby nodes configs, snapshot the
+// states, and validate the node has a clean state.
+func (h *HarnessTest) cleanupStandbyNode(hn *node.HarnessNode) {
+	// Remove connections made from this test.
+	h.removeConnectionns(hn)
+
+	// Delete all payments made from this test.
+	hn.RPC.DeleteAllPayments()
+
+	// Update the node's internal state.
+	hn.UpdateState()
+
+	// Finally, check the node is in a clean state for the following tests.
+	h.validateNodeState(hn)
+}
+
+// removeConnectionns will remove all connections made on the standby nodes
+// expect the connections between Alice and Bob.
+func (h *HarnessTest) removeConnectionns(hn *node.HarnessNode) {
+	resp := hn.RPC.ListPeers()
+	for _, peer := range resp.Peers {
+		// Skip disconnecting Alice and Bob.
+		switch peer.PubKey {
+		case h.Alice.PubKeyStr:
+			continue
+		case h.Bob.PubKeyStr:
+			continue
+		}
+
+		hn.RPC.DisconnectPeer(peer.PubKey)
+	}
+}
+
+// SetTestName set the test case name.
+func (h *HarnessTest) SetTestName(name string) {
+	h.manager.currentTestCase = name
+
+	// Overwrite the old log filename so we can create new log files.
+	for _, node := range h.manager.standbyNodes {
+		node.Cfg.LogFilenamePrefix = name
+	}
+}
+
+// NewNode creates a new node and asserts its creation. The node is guaranteed
+// to have finished its initialization and all its subservers are started.
+func (h *HarnessTest) NewNode(name string,
+	extraArgs []string) *node.HarnessNode {
+
+	node, err := h.manager.newNode(h.T, name, extraArgs, false, nil, false)
+	require.NoErrorf(h, err, "unable to create new node for %s", name)
+
+	return node
+}
+
+// Shutdown shuts down the given node and asserts that no errors occur.
+func (h *HarnessTest) Shutdown(node *node.HarnessNode) {
+	// The process may not be in a state to always shutdown immediately, so
+	// we'll retry up to a hard limit to ensure we eventually shutdown.
+	err := wait.NoError(func() error {
+		return h.manager.shutdownNode(node)
+	}, DefaultTimeout)
+	require.NoErrorf(h, err, "unable to shutdown %v", node.Name())
```

Comment:
Looks like another useful pattern for the `HarnessNode` to provide as a method: `node.WaitNoError(func() error)` that uses the `DefaultTimeout` and aborts on error.